compiler/src/iree/compiler/GlobalOptimization/ExpandVectors.cpp

```diff
@@ -0,0 +1,137 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//===--------------- ExpandVectors.cpp -----------------------------------===//
+// Expands vectors in matrix/vector operations (vecmat, matvec, batch_matvec)
+// into matrices in order to enable tiling.
+//===---------------------------------------------------------------------===//
+
+#include "iree/compiler/GlobalOptimization/PassDetail.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace GlobalOptimization {
+
+namespace {
+
+struct ExpandVectors
+    : public OpInterfaceRewritePattern<linalg::ContractionOpInterface> {
+  using OpInterfaceRewritePattern<
+      linalg::ContractionOpInterface>::OpInterfaceRewritePattern;
+  LogicalResult matchAndRewrite(linalg::ContractionOpInterface op,
+                                PatternRewriter &rewriter) const override {
+    auto linalgOp = dyn_cast<linalg::LinalgOp>(op.getOperation());
+    if (!linalgOp || !linalgOp.hasTensorSemantics())
+      return failure();
+
+    Value lhs = linalgOp.getDpsInputs()[0];
+    Value rhs = linalgOp.getDpsInputs()[1];
+
+    Value vectorIn;
+    if (op.isVecmat()) {
+      vectorIn = lhs;
+    } else if (op.isMatvec() || op.isBatchMatvec()) {
+      vectorIn = rhs;
+    } else {
+      return rewriter.notifyMatchFailure(linalgOp,
+                                         "unsupported contraction op");
+    }
+
+    auto vectorOut = linalgOp.getDpsInits()[0];
+    auto vectorInTy = llvm::dyn_cast<RankedTensorType>(vectorIn.getType());
+    auto vectorOutTy = llvm::dyn_cast<RankedTensorType>(vectorOut.getType());
+
+    if (!vectorInTy || !vectorOutTy) {
+      return failure();
+    }
+
+    Type inEType = vectorInTy.getElementType();
+    Type outEType = vectorOutTy.getElementType();
+
+    SmallVector<int64_t> expandedInDims, expandedOutDims;
+    bool isBatchMatmul = false;
+    SmallVector<ReassociationIndices> ri = {{0, 1}};
+
+    // Expand (N * NxM = M) into (1xN * NxM = 1xM)
+    if (op.isVecmat()) {
+      expandedInDims = {1, vectorInTy.getDimSize(0)};
+      expandedOutDims = {1, vectorOutTy.getDimSize(0)};
+      // Expand (NxM * M = N) into (NxM * Mx1 = Mx1)
+    } else if (op.isMatvec()) {
+      expandedInDims = {vectorInTy.getDimSize(0), 1};
+      expandedOutDims = {vectorOutTy.getDimSize(0), 1};
+      // Expand (BxNxM * BxM = BxN) into (BxNxM * BxMx1 = BxMx1)
+    } else {
+      expandedInDims = {vectorInTy.getDimSize(0), vectorInTy.getDimSize(1), 1};
+      expandedOutDims = {vectorOutTy.getDimSize(0), vectorOutTy.getDimSize(1),
+                         1};
+      ri = {{0}, {1, 2}};
+      isBatchMatmul = true;
+    }
```

Comment:
[optional] I'm not a fan of putting comments in this style. It's quite ambiguous to me. Can you move the comments to corresponding block?

```suggestion
    if (op.isVecmat()) {
      // Expand (N * NxM = M) into (1xN * NxM = 1xM)
      expandedInDims = {1, vectorInTy.getDimSize(0)};
      expandedOutDims = {1, vectorOutTy.getDimSize(0)};
    } else if (op.isMatvec()) {
      // Expand (NxM * M = N) into (NxM * Mx1 = Mx1)
      expandedInDims = {vectorInTy.getDimSize(0), 1};
      expandedOutDims = {vectorOutTy.getDimSize(0), 1};
    } else {
      // Expand (BxNxM * BxM = BxN) into (BxNxM * BxMx1 = BxMx1)
      expandedInDims = {vectorInTy.getDimSize(0), vectorInTy.getDimSize(1), 1};
      expandedOutDims = {vectorOutTy.getDimSize(0), vectorOutTy.getDimSize(1),
                         1};
      ri = {{0}, {1, 2}};
      isBatchMatmul = true;
    }
```

Reply:
I agree that this is clearer, thanks!